apis/metal3.io/v1alpha1/firmwareschema_types.go

```diff
@@ -0,0 +1,147 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// Additional data describing the firmware setting
+type SettingSchema struct {
+
+	// The type of setting.
+	// +kubebuilder:validation:Enum=Enumeration;String;Integer;Boolean;Password
+	AttributeType string `json:"attribute_type,omitempty"`
+
+	// The allowable value for an Enumeration type setting.
+	AllowableValues []string `json:"allowable_values,omitempty"`
+
+	// The lowest value for an Integer type setting.
+	LowerBound *int `json:"lower_bound,omitempty"`
+
+	// The highest value for an Integer type setting.
+	UpperBound *int `json:"upper_bound,omitempty"`
+
+	// Minimum length for a String type setting.
+	MinLength *int `json:"min_length,omitempty"`
+
+	// Maximum length for a String type setting.
+	MaxLength *int `json:"max_length,omitempty"`
+
+	// Whether or not this setting is read only.
+	ReadOnly *bool `json:"read_only,omitempty"`
+
+	// Whether or not a reset is required after changing this setting.
+	ResetRequired *bool `json:"reset_required,omitempty"`
+
+	// Whether or not this setting's value is unique to this node, e.g.
+	// a serial number.
+	Unique *bool `json:"unique,omitempty"`
+}
+
+// FirmwareSchemaSpec defines the desired state of FirmwareSchema
+type FirmwareSchemaSpec struct {
+
+	// The hardware vendor associated with this schema
+	// +optional
+	HardwareVendor string `json:"hardwareVendor,omitempty"`
+
+	// The hardware model associated with this schema
+	// +optional
+	HardwareModel string `json:"hardwareModel,omitempty"`
+
+	// Map of firmware name to schema
+	Schema map[string]SettingSchema `json:"schema" required:"true"`
+}
+
+//+kubebuilder:object:root=true
+
+// FirmwareSchema is the Schema for the firmwareschemas API
+type FirmwareSchema struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec FirmwareSchemaSpec `json:"spec,omitempty"`
+}
+
+// Check whether the setting's name and value is valid using the schema
+func (host *FirmwareSchema) CheckSettingIsValid(name string, value intstr.IntOrString, schemas map[string]SettingSchema) bool {
+
+	schema, ok := schemas[name]
+	if !ok {
+		// The setting must exist in the status
+		return false
+	}
+
+	if schema.ReadOnly != nil && *schema.ReadOnly == true {
```

Comment:
I wonder if we need any logging in this function. Otherwise debugging can be challenging.

Or can we maybe return an `error` instead and log it somewhere at the caller place?

Reply:
Good point about the debugging, I will add more granular debugging to help determine error in settings.

Reply:
Changed CheckSettingIsValid() in https://github.com/metal3-io/baremetal-operator/pull/938 to return an error which is now logged.